### PR TITLE
Convert pagination_links.jsp template to handlebars

### PR DESCRIPTION
**Pull request by PaulMorris**
This template renders the `Pages: << Prev 2 3 4 Next >>` part of the provider dashboard pages (e.g. `/cms/provider/dashboard/`, `/cms/provider/dashboard/filter`). The calculations in the JSP template are moved into ProviderDashboardController.java.

Manually tested using next and prev links as well as number links (1 2 3), and also by filtering with a query that returns 0 results, so that this template displays nothing on the page.

Issue #238 Templatize UI

### CODE REVIEW

File: psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    {{/if}}
+    {{#each nextPages }}
+    <a href="javascript:changePageNumber({{this}})">{{this}}</a>
+    {{/each}}
+    <span class="pageNumber">{{currentPage}}</span>
+    {{#each prevPages }}
+    <a href="javascript:changePageNumber({{this}})">{{this}}</a>
+    {{/each}}
+    {{#if prevPages }}
+    <a href="javascript:changePageNumber({{currentPage}} - 1)">&lt;&lt; Prev</a>
```
Comment (jasonaowen) [inline psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html]: Would you mind indenting the contents of the `#if` and `#each` blocks?

File: psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    {{/if}}
+  </div>
+  <span class="floatR">Pages: </span>
```
Comment (jasonaowen) [inline psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html]: I recognize that this is the same as the JSP version, but do you know why this span is at the end? Can it be moved to the beginning without breaking things?

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java [hunk 3]
```diff
@@ -345,6 +347,27 @@ private ModelAndView searchTickets(ProviderSearchCriteria criteria, String view)
...
+        ArrayList<Integer> nextPages = new ArrayList<Integer>();
+        for (Integer i = lastPage; i >= firstPage; i--) {
+            if (i < pageNumber) {
+                prevPages.add(i);
+            } else if (i > pageNumber) {
+                nextPages.add(i);
+            }
+        }
+        mv.addObject("prevPages", prevPages);
+        mv.addObject("nextPages", nextPages);
```
Comment (jasonaowen) [inline psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java]: This is an optional extra: would you mind extracting a helper method to add all the pagination stuff to the model? Maybe something like `private void addPaginationVariables(SearchResult<?> results, ModelAndView mv)`. I think putting this code into its own method would make it a bit easier to read both it and the `searchTickets` method.

And, as a double optional bonus, that would make it testable - I know I haven't been modeling that behavior as well as I would like, but this kind of standalone, largely-dependency-free code is a perfect candidate for adding some unit tests. I can work on that with you on Wednesday if you like, @PaulMorris !

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java [hunk 4]
```diff
@@ -345,6 +347,27 @@ private ModelAndView searchTickets(ProviderSearchCriteria criteria, String view)
                 true
         );
 
+        mv.addObject("thereArePages", totalPages > 0);
+        mv.addObject("currentPage", pageNumber);
+
+        Integer firstPage = (pageNumber + 4 > totalPages) ? totalPages - 4 : pageNumber - 2;
```
[the remark this answers is not in the document]
Comment (cecilia-donnelly) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java]: +1

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java [hunk 5]
```diff
@@ -345,6 +347,27 @@ private ModelAndView searchTickets(ProviderSearchCriteria criteria, String view)
...
+        firstPage = firstPage < 1 ? 1 : firstPage;
+
+        Integer lastPage = firstPage + 4;
+        lastPage = lastPage > totalPages ? totalPages : lastPage;
+
+        ArrayList<Integer> prevPages = new ArrayList<Integer>();
+        ArrayList<Integer> nextPages = new ArrayList<Integer>();
```
[the remark this answers is not in the document]
Comment (cecilia-donnelly) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java]: Thanks, @jasonaowen, this is good to know!

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java (hunk 3 above)
Comment (cecilia-donnelly) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java]: A hearty +1 to adding tests!

File: psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html (hunk 1 above)
Comment (PaulMorris) [reply psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html]: Will do.  (I had them indented, but then un-indented them so the resulting HTML would be indented correctly, but on second thought better to have it indented in the source for easier reading.)

File: psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html (hunk 2 above)
Comment (PaulMorris) [reply psm-app/cms-web/WebContent/templates/includes/pagination_links.template.html]: The various elements here are all floated right in CSS so that they have to be in reverse order in the source HTML to come out correctly on the screen.  Although this is counter-intuitive, my first inclination was to leave it as is rather than take more time on it, but I'm up for changing it to a more intuitive order if we think it's worth it.  (Hmmm, this might also be an accessibility issue as the items would be backwards in the HTML.)

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (PaulMorris) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java]: OK, I'll clean this up as suggested.  This previously existing logic could definitely be clearer.

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java (hunk 3 above)
Comment (PaulMorris) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java]: Sounds good!

**Comment by PaulMorris** [comment]
This revision addresses all of @jasonaowen 's comments (including reordering the elements and removing the float right) except for adding a unit test.  Perhaps the test can be added in a follow-up PR?  I've separated out the commits to make it easier to review the changes, but I plan on squashing them into one commit before merging this.

File: psm-app/cms-web/src/test/groovy/gov/medicaid/controllers/ProviderDashboardControllerTest.groovy [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        when:
+        controller.addPaginationLinks(results, mv)
+
+        then:
+        def model = mv.getModel()
+        model["prevPages"] == prevPages
+        model["nextPages"] == nextPages
+        model["thereArePages"] == thereArePages
+        model["currentPage"] == pageNumber
+        model["nextPages"].size() + model["prevPages"].size() <= controller.MAX_PAGE_LINKS_TO_SHOW
```
Comment (jasonaowen) [inline psm-app/cms-web/src/test/groovy/gov/medicaid/controllers/ProviderDashboardControllerTest.groovy]: Sorry, I just noticed `controller.MAX_PAGE_LINKS_TO_SHOW` - we should avoid referencing static members of an instance. Can you please change this to a static import? I think you can do that by replacing `controller` with `ProviderDashboardController`, and then clicking on it and hitting alt-enter to import static. Sorry for not catching this while pairing!

File: psm-app/cms-web/src/test/groovy/gov/medicaid/controllers/ProviderDashboardControllerTest.groovy [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        2          | 25       | 26         || true          | [1]          | []
+        2          | 10       | 30         || true          | [1]          | [3]
+        1          | 0        | 15         || true          | []           | []
+        5          | 10       | 50         || true          | [1, 2, 3, 4] | []
+        5          | 10       | 60         || true          | [2, 3, 4]    | [6]
+        4          | 10       | 60         || true          | [2, 3]       | [5, 6]
+        3          | 10       | 60         || true          | [2]          | [4, 5, 6]
+        3          | 10       | 70         || true          | [1, 2]       | [4, 5]
+        1          | 10       | 70         || true          | []           | [2, 3, 4, 5]
+        2          | 10       | 70         || true          | [1]          | [3, 4, 5]
```
Comment (jasonaowen) [inline psm-app/cms-web/src/test/groovy/gov/medicaid/controllers/ProviderDashboardControllerTest.groovy]: These tests are really useful for capturing how this works!

@cecilia-donnelly  :eyes:

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java [hunk 8]
```diff
@@ -325,28 +332,74 @@ private ModelAndView searchTickets(ProviderSearchCriteria criteria, String view)
...
+
+    void addPaginationLinks(SearchResult<?> results, ModelAndView mv) {
+        int pageNumber = results.getPageNumber();
+        int totalPages = results.getTotalPages();
+
+        mv.addObject("thereArePages", totalPages > 0);
+        mv.addObject("currentPage", pageNumber);
+
+        int firstPage = getFirstPage(pageNumber, totalPages);
+        int lastPage = min(firstPage + MAX_PAGE_LINKS_TO_SHOW, totalPages);
```
Comment (cecilia-donnelly) [inline psm-app/cms-web/src/main/java/gov/medicaid/controllers/ProviderDashboardController.java]: This is way easier to read.  Thank you both!

File: psm-app/cms-web/src/test/groovy/gov/medicaid/controllers/ProviderDashboardControllerTest.groovy (hunk 7 above)
Comment (cecilia-donnelly) [reply psm-app/cms-web/src/test/groovy/gov/medicaid/controllers/ProviderDashboardControllerTest.groovy]: They really are!

**Comment by PaulMorris** [comment]
This revision fixes how we were accessing MAX_PAGE_LINKS_TO_SHOW in the tests file.  Thanks @jasonaowen for catching that.  @cecilia-donnelly we were running the tests from IntelliJ, so there may be more to do to get them to run with the `./gradlew` command.

**Comment by cecilia-donnelly** [comment]
@PaulMorris can you write up how to get them to run with gradle?  In particular, I want to make sure that they're being run correctly by our CI, which uses gradle commands.

**Comment by jasonaowen** [comment]
These tests are part of `cms-web`, so you can run them with `./gradlew cms-web:test`.

Checking that they're running in CI is a great idea, which I hadn't thought about! [Here are the lines from the build log](https://travis-ci.org/OpenTechStrategies/psm/builds/299715285#L6088-L6105). We're running those tests as part of [`./gradlew test` in Travis](https://travis-ci.org/OpenTechStrategies/psm/builds/299715285#L5726), which includes the integration tests.

**Comment by jasonaowen** [comment]
Hm, I seem to have found an edge case. Logged in as provider with 67 applications in various states (pending, denied, approved, draft), on the dashboard I switched to "Show: All", and the behavior is wrong on both master and this PR, but differently...

Both go to page 0 (set as a query parameter), but then the pagination does different things...
- Master shows no pages: master
- This PR shows that we're on page 0, and there is a page 1: pr

**Comment by jasonaowen** [comment]
@PaulMorris you've studied this code more carefully than I have; the two options I see are to make it so switching to "All" makes it be on page 1 instead of page 0, or to make the pagination handle being on page 0 more gracefully. Is there another? Which do you suggest?

**Comment by jasonaowen** [comment]
Also, I think I would prefer to fix this in this PR, as we are changing the behavior to something that feels less good to me. `Pages:` is weird, but `Pages: 0 1 Next >>` is even weirder!

**Comment by PaulMorris** [comment]
Good catch @jasonaowen !  (I don't have that many enrollments in my database and didn't see this.)  Let me take another look at the code and see how to best fix it.

**Comment by PaulMorris** [comment]
@jasonaowen  Letting "All" set the page number to 1 instead of 0 did the trick!  All it took was a simple change in script.js.  In the "All" case it now shows "Pages: 1".  I did not see any negative effects of this change.

**Comment by cecilia-donnelly** [comment]
@PaulMorris: While logged in as an admin, I chose "all" for the "pending" tab, which had three enrollments.  Instead of Pages: 1 it didn't show anything at all.  I think this is a fine UI choice, but wanted to raise it in case it's unexpected.

**Comment by PaulMorris** [comment]
@cecilia-donnelly Thanks for testing.  I should have specified that these template changes only affect the provider part of the site, so you won't see them when logged in as admin.  Admin has its own separate pagination templates (e.g. admin/includes/top-pagination-section.jsp).  (Clearly a good opportunity for some consolidation.)

**Comment by PaulMorris** [comment]
Also, these templates are only used for the following pages/urls in the provider part of the site:
/provider/dashboard/
/provider/dashboard/list/
/provider/dashboard/list/filter

File: psm-app/cms-web/WebContent/js/script.js [hunk 9]
```diff
@@ -2149,14 +2149,8 @@ function changeSort(idx) {
...
-  if (size == 0) {
-    $('#paginationForm input[name=pageSize]').val(0);
-    $('#paginationForm input[name=pageNumber]').val(0);
-  } else {
-    $('#paginationForm input[name=pageSize]').val(size);
-    $('#paginationForm input[name=pageNumber]').val(1);
-  }
-
+  $('#paginationForm input[name=pageSize]').val(size);
+  $('#paginationForm input[name=pageNumber]').val(1);
```
Comment (jasonaowen) [inline psm-app/cms-web/WebContent/js/script.js]: Nice!
